### PR TITLE
SCP-1861 Marlowe Playground Client: Implement holes in blockly

### PR DESCRIPTION
**Pull request by hrajchert**
This PR implements blockly holes, which is a way to suggest to the user what blocks could fit into the selected block.

I had to upgrade blockly to the latest version (and while I was at it modify the source to the official one) as it was necessary for defining the Toolbox as JSON. Then I re-implement the toolbox from XML to Json and finally added the mechanism to modify the Toolbox with the hole information.

As a "weird" usability issue, the names of the toolbox categories correspond to the name of the holes rather than the name of the block. But trying to change that broke the CodeToBlocklyToCode quick check, as the names of the holes is tightly coupled to the block definition, and the quick check verifies that the names of the holes are preserved.

The code is deployed here: https://hernan.marlowe.iohkdev.io/

Screen Shot 2021-03-02 at 12 12 51
Screen Shot 2021-03-02 at 12 13 03

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [ ] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [x] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [x] Someone approved it
- [x] Commits have useful messages
- [x] Review clarifications made it into the code
- [x] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: marlowe-playground-client/src/BlocklyComponent/Types.purs [hunk 1]
```diff
@@ -43,16 +44,18 @@ data Query a
...
-  = Inject String (Array BlockDefinition)
+  = Inject String (Array BlockDefinition) Toolbox
   | SetData Unit
   | BlocklyEvent BT.BlocklyEvent
   | ResizeWorkspace
   | VisibilityChanged Boolean
 
 data Message
   = CodeChange
+  | BlockSelection (Maybe ({ blockId :: String, blockType :: String }))
```
Comment (palas) [inline marlowe-playground-client/src/BlocklyComponent/Types.purs:58]: I imagine these parenthesis around the curly brackets are not necessary

File: marlowe-playground-client/src/Marlowe/Blockly.purs [hunk 2]
```diff
@@ -1000,6 +1006,79 @@ toDefinition blockType@(ValueType UseValueValueType) =
...
+payeeCategory = category "Payee" payeeColour $ map (leaf <<< show) payeeTypes
+
+partyCategory :: Category
+partyCategory = category "Party" partyColour $ map (leaf <<< show) partyTypes
+
+tokenCategory :: Category
+tokenCategory = category "Token" tokenColour $ map (leaf <<< show) tokenTypes
+
+boundsCategory :: Category
+boundsCategory = category "Bounds" boundsColour $ [ leaf $ show BoundsType ]
```
Comment (palas) [inline marlowe-playground-client/src/Marlowe/Blockly.purs:1035]: The first dollar symbol here is not necessary

File: marlowe-playground-client/src/Marlowe/Blockly.purs [hunk 3]
```diff
@@ -1000,6 +1006,79 @@ toDefinition blockType@(ValueType UseValueValueType) =
...
+
+-- This map ties a block definition "check" with a Category that can be displayed in the Toolbar
+-- as part of the Blockly holes functionality.
+-- TODO: We should probably stop relying on `Show` for check and use a NewType to avoid having any string
+typeMap :: Map String Category
+typeMap =
+  Map.fromFoldable
+    $ [ Tuple (show BaseContractType) contractsCategory
```
Comment (palas) [inline marlowe-playground-client/src/Marlowe/Blockly.purs:1043]: Same here :)

File: marlowe-playground-client/src/Blockly/Internal.purs [hunk 4]
```diff
@@ -231,6 +241,13 @@ data Arg
   | DummyLeft
   | DummyCentre
 
+argType :: Arg -> Maybe ({ name :: String, check :: String })
+argType (Value { name, check }) = Just $ { name, check }
+
+argType (Statement { name, check }) = Just $ { name, check }
```
Comment (palas) [inline marlowe-playground-client/src/Blockly/Internal.purs:247]: Also here in 245 and 247

File: marlowe-playground-client/src/BlocklyComponent/State.purs [hunk 5]
```diff
@@ -152,6 +163,16 @@ handleAction (Inject rootBlockName blockDefinitions) = do
...
 
+handleAction (BlocklyEvent (BT.Select event)) = case newElementId event of
+  Nothing -> raise $ BlockSelection Nothing
+  Just blockId -> do
+    void
+      $ runMaybeT do
+          blocklyState <- MaybeT $ use _blocklyState
+          block <- MaybeT $ liftEffect $ getBlockById blocklyState.workspace blockId
+          blockType <- MaybeT $ map pure $ liftEffect $ getBlockType block
+          MaybeT $ map pure $ raise $ BlockSelection $ Just $ { blockId, blockType }
```
Comment (palas) [inline marlowe-playground-client/src/BlocklyComponent/State.purs:174]: Also here

**Review by palas — APPROVED** [review]
Nice! Looks good. I just wrote a couple of lints

File: marlowe-playground-client/src/BlocklyComponent/Types.purs (hunk 1 above)
Comment (hrajchert) [reply marlowe-playground-client/src/BlocklyComponent/Types.purs:58]: done

File: marlowe-playground-client/src/Marlowe/Blockly.purs (hunk 3 above)
Comment (hrajchert) [reply marlowe-playground-client/src/Marlowe/Blockly.purs:1043]: done

**Review by merivale — APPROVED** [review]
I don't know blockly, but looks like some nice housekeeping alongside a great new feature. 👍 Once hydra's happy, I'm happy.

To be clear, is the reason for Argonaut that blockly expects the JSON in a particular format, that's different from the foreign-generic default?

**Comment by hrajchert** [comment]
Yes, basically thats the reason. After doing this PR I saw that we can change the shape with foreign-generic using stuff like [this](https://github.com/input-output-hk/plutus/blob/master/web-common-marlowe/src/Marlowe/Extended.purs#L201). But I need to check how it would work with optional fields (the catMaybe parts).

In any case, argonaut-core was already included in the dependencies and it supposedly ready to be used by servant-purescript. We should check if it's worth moving code to foreing-generic/argonaut ecosystem, or if its fine to have a mix. It would also be nice to see what direction is the community heading to.

**Pull request merged**